smarts/core/vehicle_index.py

```diff
@@ -76,22 +95,31 @@ def __sub__(self, other: "VehicleIndex") -> "VehicleIndex":
         vehicle_ids = set(self._controlled_by["vehicle_id"]) - set(
             other._controlled_by["vehicle_id"]
         )
+
+        vehicle_ids = [self._2id_to_id[id_] for id_ in vehicle_ids]
         return self._subset(vehicle_ids)
 
     def __and__(self, other: "VehicleIndex") -> "VehicleIndex":
         vehicle_ids = set(self._controlled_by["vehicle_id"]) & set(
             other._controlled_by["vehicle_id"]
         )
+
+        vehicle_ids = [self._2id_to_id[id_] for id_ in vehicle_ids]
         return self._subset(vehicle_ids)
 
     def _subset(self, vehicle_ids):
-        index = VehicleIndex()
-        assert self.vehicle_ids.issuperset(vehicle_ids)
+        assert self.vehicle_ids().issuperset(
+            vehicle_ids
+        ), f"{', '.join(list(self.vehicle_ids())[:3])} ⊅ {', '.join(list(vehicle_ids)[:3])}"
```

Comment:
I think the eventual optimization of this vehicle index will be to lower the number of collection allocations.